Change FakeTensor constructor to use `_make_subclass`

eellison opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #78524
* #78523
* #78522
* #78517
* #78270
* #78090
* #78516
* #77972
* #77971
* __->__ #77970
* #77969

Add an overload for `new` to handle setting the custom strides/custom device property.

Commit:
Extend __new__ on subclasses to set custom_device and custom_strides

[ghstack-poisoned]

Commit:
Update on "Extend __new__ on subclasses to set custom_device and custom_strides"


Add an overload for `new` to handle setting the custom strides/custom device property.

[ghstack-poisoned]

Commit:
Update on "Extend __new__ on subclasses to set custom_device and custom_strides"


Add an overload for `new` to handle setting the custom strides/custom device property.

[ghstack-poisoned]

Commit:
Update on "Extend __new__ on subclasses to set custom_device and custom_strides"


Add an overload for `new` to handle setting the custom strides/custom device property.

[ghstack-poisoned]

Commit:
Update on "Extend __new__ on subclasses to set custom_device and custom_strides"


Add an overload for `new` to handle setting the custom strides/custom device property.

[ghstack-poisoned]

ezyang commented on torch/_subclasses/fake_tensor.py:
```diff
@@ -44,16 +44,22 @@ def from_tensor(t):
-    @property
-    def device(self):
-        return self.fake_device
+    # TODO: resolve error in default __repr__
```
What's the error?

eellison replied:
Haven't really looked into it yet, but this is the callstack and the `func` that is invoked:
```
aten.add.Tensor
prim.device.default
prim.device.default
prim.device.default
prim.device.default
prim.device.default
prim.device.default
prim.device.default
prim.device.default
prim.device.default
prim.device.default
prim.device.default
prim.device.default
prim.device.default
prim.device.default
prim.device.default
prim.device.default
prim.device.default
prim.device.default
prim.device.default
prim.device.default
prim.device.default
prim.device.default
aten._reshape_alias.default
prim.device.default
prim.device.default
prim.device.default
aten.abs.default
prim.device.default
prim.device.default
prim.device.default
F
======================================================================
FAIL: test_zero_dim (__main__.FakeTensorTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/scratch/eellison/pytorch/test/test_fake_tensor.py", line 31, in test_zero_dim
    print(x)
  File "/private/home/eellison/anaconda3/lib/python3.8/site-packages/torch/_tensor.py", line 338, in __repr__
    return torch._tensor_str._str(self, tensor_contents=tensor_contents)
  File "/private/home/eellison/anaconda3/lib/python3.8/site-packages/torch/_tensor_str.py", line 481, in _str
    return _str_intern(self, tensor_contents=tensor_contents)
  File "/private/home/eellison/anaconda3/lib/python3.8/site-packages/torch/_tensor_str.py", line 447, in _str_intern
    tensor_str = _tensor_str(self, indent)
  File "/private/home/eellison/anaconda3/lib/python3.8/site-packages/torch/_tensor_str.py", line 270, in _tensor_str
    formatter = _Formatter(get_summarized_data(self) if summarize else self)
  File "/private/home/eellison/anaconda3/lib/python3.8/site-packages/torch/_tensor_str.py", line 103, in __init__
    nonzero_finite_vals = torch.masked_select(tensor_view, torch.isfinite(tensor_view) & tensor_view.ne(0))
  File "/private/home/eellison/anaconda3/lib/python3.8/site-packages/torch/_subclasses/fake_tensor.py", line 98, in __torch_dispatch__
    return tree_map(partial(wrap, device=common_device), r)
  File "/private/home/eellison/anaconda3/lib/python3.8/site-packages/torch/utils/_pytree.py", line 179, in tree_map
    return tree_unflatten([fn(i) for i in flat_args], spec)
  File "/private/home/eellison/anaconda3/lib/python3.8/site-packages/torch/utils/_pytree.py", line 179, in <listcomp>
    return tree_unflatten([fn(i) for i in flat_args], spec)
  File "/private/home/eellison/anaconda3/lib/python3.8/site-packages/torch/_subclasses/fake_tensor.py", line 72, in wrap
    return FakeTensor(e, device)
  File "/private/home/eellison/anaconda3/lib/python3.8/site-packages/torch/_subclasses/fake_tensor.py", line 39, in __init__
    assert elem.device.type == "meta"
AssertionError
```

ezyang replied:
Oh you need to add a special case for fake tensor in printing the same way we have a short circuit for meta printing. The meta printing short circuit no longer works because your device isn't meta anymore

ezyang commented:
I'm a bit iffy on this. Why not extend `_make_subclass` and then call it directly from `__new__`?

eellison commented:
The chain of calls is 
`__new__` -> [THPVariable_pynew](https://github.com/pytorch/pytorch/blob/master/torch/csrc/autograd/python_variable.cpp#L1465) -> [base_tensor_ctor](https://github.com/pytorch/pytorch/blob/master/torch/csrc/utils/tensor_new.cpp#L573) -> [legacy_tensor_generic_ctor_new](https://github.com/pytorch/pytorch/blob/master/torch/csrc/utils/tensor_new.cpp#L491). 

`legacy_tensor_generic_ctor_new` is what does the parsing of `new` schemas and where I added the logic here. Once it returns the new Tensor to `THPVariable_pynew` then [THPVariable_NewWithVar](https://github.com/pytorch/pytorch/blob/master/torch/csrc/autograd/python_variable.cpp#L1468) does the wrapping of the new tensor to its potential Tensor subclass. 

[make_subclass](https://github.com/pytorch/pytorch/blob/master/torch/csrc/autograd/python_variable.cpp#L536) does the parsing of schemas (`_make_subclass(PyObject* cls, Tensor data, bool require_grad=False, *, bool dispatch_strides=False`) , the creation of the new Tensor, and the wrapping of its new tensor to its potential Tensor subclass. 

So it's really only 
```
    if (r.toBool(1)) {
      alias.unsafeGetTensorImpl()->set_sizes_strides_policy(c10::TensorImpl::SizesStridesPolicy::CustomStrides);
    }
    if (r.toBool(2)) {
      alias.unsafeGetTensorImpl()->set_custom_device(true);
    }
    return alias;
```
that could be shared without a lot of refactoring, since `_make_subclass` both does parsing of schemas and the wrapping of tensors to its subclasses...

Unless there's some easy refactoring I'm missing or larger refactoring I should do doesn't seem like a lot to share ?

Commit:
Update on "Extend __new__ on subclasses to set custom_device and custom_strides"


Add an overload for `new` to handle setting the custom strides/custom device property.

[ghstack-poisoned]

ezyang commented:
It's not about refactoring to share; it's related to the earlier question which is why you needed to use `BaseTensor.__new__` as opposed to defining it directly and calling `_make_subclass` inside `__new__`, so you wouldn't exercise the `__new__` codepath at all.

BTW, the underlying motivation here is that `__new__` is exercisable by anyone who does a `Tensor(...)` call, so I'd like to avoid sprouting API surface here.

Commit:
Update on "Extend __new__ on subclasses to set custom_device and custom_strides"


Add an overload for `new` to handle setting the custom strides/custom device property.

[ghstack-poisoned]

eellison commented:
From @samdow 
> So typically if you pass requires_grad to a tensor constructor, that means that you want the subclass that you're creating to requires gradient, not the tensor that it's wrapping. It looks like from this https://github.com/pytorch/pytorch/pull/73727, it accepts a single tensor, so super().__new__() calls that and if the tensor requires gradient, it will wrap it like normal and won't put requires_grad on the wrapper tensor

as to why 
```
    def __new__(cls, elem, *, requires_grad=None):
        if requires_grad is None:
            return super().__new__(cls, elem)
        else:
            return cls._make_subclass(cls, elem, requires_grad)
```
is required

but I updated to just call `_make_subclass` and it seems to work for now. Maybe something that will need to be revisited later

Commit:
Update on "Extend __new__ on subclasses to set custom_device and custom_strides"


Add an overload for `new` to handle setting the custom strides/custom device property.

[ghstack-poisoned]

Chillee reviewed (approved):
LGTM.

Commit:
Update on "Change FakeTensor constructor to use `_make_subclass`"


Add an overload for `new` to handle setting the custom strides/custom device property.

[ghstack-poisoned]

eellison commented:
@eellison has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D36618466).

Commit:
Update on "Change FakeTensor constructor to use `_make_subclass`"


Add an overload for `new` to handle setting the custom strides/custom device property.

Differential Revision: [D36618466](https://our.internmc.facebook.com/intern/diff/D36618466)

[ghstack-poisoned]

Commit:
Update on "Change FakeTensor constructor to use `_make_subclass`"


Add an overload for `new` to handle setting the custom strides/custom device property.

Differential Revision: [D36618466](https://our.internmc.facebook.com/intern/diff/D36618466)

[ghstack-poisoned]

Commit:
Update on "Change FakeTensor constructor to use `_make_subclass`"


Add an overload for `new` to handle setting the custom strides/custom device property.

Differential Revision: [D36618466](https://our.internmc.facebook.com/intern/diff/D36618466)

[ghstack-poisoned]

Commit:
Update on "Change FakeTensor constructor to use `_make_subclass`"


Add an overload for `new` to handle setting the custom strides/custom device property.

Differential Revision: [D36618466](https://our.internmc.facebook.com/intern/diff/D36618466)

[ghstack-poisoned]

Commit:
Update on "Change FakeTensor constructor to use `_make_subclass`"


Add an overload for `new` to handle setting the custom strides/custom device property.

Differential Revision: [D36618466](https://our.internmc.facebook.com/intern/diff/D36618466)

[ghstack-poisoned]

Commit:
Update on "Change FakeTensor constructor to use `_make_subclass`"


Add an overload for `new` to handle setting the custom strides/custom device property.



[ghstack-poisoned]

eellison commented:
@pytorchbot merge this please
